### PR TITLE
Added Route Guard and Persist Logged In User

### PR DESCRIPTION
**Pull request by vipulsnp**
This PR solves issue #87 #86 

**What does this PR do?**
If no user is logged in and tries to access any page other than home, he is redirected to home. 
Logged In user is persisted until explicit logout.
**Any background context you want to provide?**
If a logged in user tries to access homepage, he is redirected to feed page.
Explicit logout is required.

**Screenshots?**

### CODE REVIEW

File: pages/_app.jsx [hunk 1]
```diff
@@ -9,6 +10,25 @@ function MyApp({ Component, pageProps }) {
   /* eslint-disable-next-line no-unused-vars */
   const [User, setUser] = useState(null);
 
+  useEffect(()=>{
+    let result = localStorage.getItem('User');
+    if (result !== null)
+    {
+      result = JSON.parse(result);
```
Comment (irshadjsr21) [inline pages/_app.jsx:17]: `JSON.parse` can throw an error if `result` is not a valid JSON. So please enclose it in `try catch` block.

File: pages/_app.jsx [hunk 2]
```diff
@@ -9,6 +10,25 @@ function MyApp({ Component, pageProps }) {
...
+      setUser({
+        name: result.Name,
```
Comment (irshadjsr21) [inline pages/_app.jsx:19]: Check if the name and email are present before setting the user.

File: src/components/FirebaseAuth.js [hunk 3]
```diff
@@ -7,6 +7,7 @@ export async function GoogleSignIn() {
   */
 
   return firebase.auth().signInWithPopup(provider).then((result)=>{
+    localStorage.setItem('User',JSON.stringify({Name : result.user.displayName, Email: result.user.email}));
```
Comment (irshadjsr21) [inline src/components/FirebaseAuth.js:10]: store the `userid` also.

File: pages/_app.jsx [hunk 4]
```diff
@@ -9,6 +10,25 @@ function MyApp({ Component, pageProps }) {
   /* eslint-disable-next-line no-unused-vars */
   const [User, setUser] = useState(null);
 
+  useEffect(()=>{
+    let result = localStorage.getItem('User');
```
Comment (irshadjsr21) [inline pages/_app.jsx:14]: Use lowercase for localStorage key. Also, encrypt the JSON data before storing and decrypt it after getting the item.

File: pages/_app.jsx (hunk 2 above)
Comment (vipulsnp) [reply pages/_app.jsx]: No need of Checking Since it will be accessed via secured token

File: pages/_app.jsx (hunk 1 above)
Comment (vipulsnp) [reply pages/_app.jsx]: No need of Checking Since it will be accessed via secured token and the returned object will always be a valid json

Comment (irshadjsr21) [reply pages/_app.jsx]: We cannot predict that. It's better to have a fallback.

File: pages/_app.jsx (hunk 2 above)
Comment (irshadjsr21) [reply pages/_app.jsx]: Same here as well. It's always better to check.

File: src/components/FirebaseAuth.js [hunk 5]
```diff
@@ -25,9 +33,33 @@ export async function GithubSignIn() {
...
   });
+}
+
+export async function getCurrentUser() {
+  return firebase.auth().currentUser;
+}
+
+export async function verifySecuredToken(token) {
+
+  return jwt.verify(token, process.env.NEXT_PUBLIC_SECURE_TOKEN_ACCESS_KEY, (err) => {
```
Comment (irshadjsr21) [inline src/components/FirebaseAuth.js:50]: `jwt.verify` also returns the decoded data. We should return that if the token is valid. We don't need a function to verify and a separate function to getUserData. We can use just one function that verifies and returns the data.

File: pages/_app.jsx [hunk 6]
```diff
@@ -1,14 +1,48 @@
...
+      async function updation() {
+        const verificationResult = await FirebaseAuth.verifySecuredToken(token);
+        if(verificationResult === true)
+         {
+           const userData = await FirebaseAuth.getUserData(token);
+          if (userData === (null || undefined) || userData.name === (null || undefined) || userData.email === (null || undefined))
+           Router.replace('/');
+           setUser({
+             name:userData.name,
+             email:userData.email
```
Comment (irshadjsr21) [inline pages/_app.jsx:26]: Set the user id as well. Also set userid while logging in.

**Comment by irshadjsr21** [comment]
Everything seems fine. Just make these changes and we're good to go.

Nice work. This was a huge step forward.

**Pull request merged**